table/block_based/block_based_table_reader.cc

```diff
@@ -3214,6 +3355,51 @@ void BlockBasedTableIterator<TBlockIter, TValue>::CheckOutOfBound() {
   }
 }
 
+template <class TBlockIter, typename TValue>
+std::string BlockBasedTableIterator<
+    TBlockIter, TValue>::tombstone_internal_start_key() const {
+  assert(range_tombstone_endpoint_.is_valid() &&
+         range_tombstone_endpoint_.dir() ==
+             RangeDelPositioningMode::kBackwardTraversal);
+  std::string internal_key;
+  if (range_tombstone_endpoint_.endpoint() == nullptr) {
+    // Make sure the key is before the file as nullptr indicates the range
+    // tombstone covers everything to the left.
+    AppendInternalKey(&internal_key, {file_meta_->smallest.user_key(),
+                                      kMaxSequenceNumber, kTypeValue});
+  } else {
+    AppendInternalKey(&internal_key,
+                      {range_tombstone_endpoint_.endpoint()->user_key,
+                       range_tombstone_endpoint_.seq(), kTypeValue});
+  }
+  return internal_key;
+}
+
+template <class TBlockIter, typename TValue>
+std::string BlockBasedTableIterator<
+    TBlockIter, TValue>::tombstone_internal_end_key() const {
+  assert(range_tombstone_endpoint_.is_valid() &&
+         range_tombstone_endpoint_.dir() ==
+             RangeDelPositioningMode::kForwardTraversal);
+  std::string internal_key;
+  if (range_tombstone_endpoint_.endpoint() == nullptr) {
+    // Make sure the key is after the file as nullptr indicates the range
+    // tombstone covers everything to the right.
+    AppendInternalKey(&internal_key,
```

Comment:
Interesting point. I do not remember why I changed it. My feeling is they're both safe since, IIRC, this is only used for seeking the iterator. If it lands earlier than it needs to, whatever keys it sees should still be covered by the normal `ShouldDelete()` check in the higher-level iterator.

On a related note, perhaps we should find a different way to communicate to the logic that seeks to the end of range tombstones that the range tombstone extends past the end of the file. Then instead of seeking to a forged key we can just invalidate the iterator.